toqito/measurements/pretty_good_measurement.py

```diff
@@ -0,0 +1,67 @@
+"""Compute the set of pretty good measurements from an ensemble."""
+import numpy as np
+import scipy
+
+from toqito.matrix_ops import vector_to_density_matrix
+
+
+def pretty_good_measurement(states: list[np.ndarray], probs: list[float] | None = None) -> list[np.ndarray]:
+    r"""Return the set of pretty good measurements from a set of vectors and corresponding probabilities.
+
+    This computes the "pretty good measurement" as initially defined in :cite:`Hughston_1993_Complete`.
+
+    See Also
+    ========
+    pretty_bad_measurement
+
+    Examples
+    ========
+    Consider the collection of trine states.
+
+    .. math::
+        u_1 = |0\rangle, \quad
+        u_1 = -\frac{1}{2}\left(|0\rangle + \sqrt{3}|1\rangle\right), \quad \text{and} \quad
+        u_2 = -\frac{1}{2}\left(|0\rangle - \sqrt{3}|1\rangle\right).
+
+    >>> from toqito.states import trine
+    >>> from toqito.measurements import pretty_good_measurement
+    >>>
+    >>> states = trine()
+    >>> probs = [1 / 3, 1 / 3, 1 / 3]
+    >>> pgm = pretty_good_measurement(states, probs)
+    >>> pgm
+    [array([[0.66666667, 0.        ],
+           [0.        , 0.        ]]), array([[0.16666667, 0.28867513],
+           [0.28867513, 0.5       ]]), array([[ 0.16666667, -0.28867513],
+           [-0.28867513,  0.5       ]])]
+
+    References
+    ==========
+        .. bibliography::
+            :filter: docname in docnames
+
+
+    :raises ValueError: If number of vectors does not match number of probabilities.
+    :raises ValueError: If probabilities do not sum to 1.
+    :param states: A collection of either states provided as either vectors or density matrices.
+    :param probs: A set of probabilities.
+
+    """
+    n = len(states)
+
+    # If not probabilities are explicitly given, assume a uniform distribution.
+    if probs is None:
+        probs = n * [1 / n]
```

Comment:
Are you assuming the input ensemble is always normalized? If not, there could be an additional check before lines 53-54 because assuming a uniform distribution by default might not work in the non-normalized scenario when the input probs are not provided.

Reply:
Hmm. I'm assuming that the probability vector is normalized (if provided). I check to ensure that the vector is a valid probability vector (i.e., it sums to 1). If such a vector is not provided, I decide to use a uniform distribution as a default (which is the section that you've highlighted here). 

Maybe I'm missing the broader point here though of what you're trying to convey.

Reply:
I am talking about a case when the provided input vector ensemble is not normalized such that it might not agree with a uniform probability distribution. 

For example, the function is provided a non-normalized input ensemble of 4 vectors but we skip providing an input probability distribution that would also normalize the input vector ensemble: `1/3, 1/6, 1/6, 1/3` is different from `1/4, 1/4, 1,4, 1,4`. 

I think it might be better to use the square root of the inner product of each vector in the non-normalized ensemble with itself to assign a probability distribution after the highlighted lines. Hopefully, this makes sense?

Reply:
Perhaps, a better example is the trine state example from both docstrings. 

![image](https://github.com/vprusso/toqito/assets/66048318/4b15033b-b7bd-4613-af0d-be41f31e18ae)
 What if the input is provided as `u_1, u_2 (without -1/2), u_3 (without -1/2)`. It won't make sense to use a uniform probability distribution `1/3, 1/3, 1/3` in this case.

Reply:
Hmm, but I think there might be a distinction here, right? The probabilities in this case are intended to be treated as "the probability with which the state is selected from the set of states" (so, the ensemble is a tuple--(p_i, rho_i) where the state is a density operator and the p_i is a probability that corresponds to the probability that rho_i is selected). So, IOW, the 1/2 and -1/2 here in this case aren't necessarily relevant to the probability vector object (they are only relevant insofar as they are used to define the actual state).

Does that make sense?

Reply:
Actually, no.